### ISSUE TITLE
Child container disposes parent container singletons

### ISSUE THREAD

**Issue by publee** [dadhi/DryIoc#338]
I want to reuse singletons in child container, but dispose them only when parent is disposed. There should be some setting to tell child container that it doesn't own the sigletons, but I can't find any...

```
[Test]
public void Test1()
{
    var parent = new Container(rules => rules.WithConcreteTypeDynamicRegistrations());
    parent.Register<IService, Service>(Reuse.Singleton);
    var service = parent.Resolve<IService>();

    var child = parent.WithRegistrationsCopy();

    var service2 = child.Resolve<IService>();
    Assert.AreEqual(service, service2);

    child.Dispose();
    Assert.IsFalse(service.IsDisposed); //child container disposed parent singleton!!!

    parent.Dispose();
    Assert.IsTrue(service.IsDisposed);
}

public interface IService
{
    bool IsDisposed { get; }
}

class Service : IService, IDisposable
{
    public void Dispose() => IsDisposed = true;
    public bool IsDisposed { get; private set; }
}

```

**Comment by dadhi** [comment]
@publee 

Hope [this doc](https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/KindsOfChildContainer.md#the-child-container-for-each-test-disposed-at-end-of-the-test-without-disposing-the-parent) will help?

I would also suggest posting the questions to the StackOverflow tagging with the **dryioc** tag.

**Comment by publee** [comment]
I've read the doc and tried all the variants, but still with no luck. I thought it is a bug, so I raised it as issue.
Here is the question on stackoveflow, lets hope someone will help: https://stackoverflow.com/questions/64406611/how-to-prevent-dryioc-to-dispose-singletons-on-child-container-dispose
Thank you.

**Comment by dadhi** [comment]
Hmm, ok. I will look then

**Comment by dadhi** [comment]
@publee Are you planning to register something to the child?

**Comment by dadhi** [comment]
@publee Let's modify your example slightly by opening the scope for the child.

```cs
var parent = new Container(rules => rules.WithConcreteTypeDynamicRegistrations());
    parent.Register<IService, Service>(Reuse.Singleton);
    var service = parent.Resolve<IService>();

    var child = parent.WithRegistrationsCopy()
        .OpenScope(); // Here!

    var service2 = child.Resolve<IService>();
    Assert.AreEqual(service, service2);

    child.Dispose();
    Assert.IsFalse(service.IsDisposed);

    parent.Dispose();
    Assert.IsTrue(service.IsDisposed);
```
Test passes (I suppose).

So the remaining part is how to bridge this with your usecase. Try to make the test fail again.

**Comment by publee** [comment]
I have IGlobalSettingsProvider in main app - main container
I have ILayoutSettingProvider in registered in child container - per window
When window is created, child container is created, ILayoutSettingProvider registered
All services resolved inside window needs to get ILayoutSettingProvider instead of IGlobalSettingsProvider 
When window is closed, child container is disposed and ILayoutSettingProvider is disposed

**Comment by publee** [comment]
Wow. Your fix is so simple and so clever! If the disposing of child container is causing trouble, just don't dispose it :)
This will definitelly work, because I am creating per container scope anyway.

**Comment by dadhi** [comment]
@publee I see, what you described may be modeled as singleton (global settings) and scopes (windows settings). I assume `IGlobalSettingsProvider` and `ILayoutSettingProvider` implement one service interface?

You may register `ILayoutSettingProvider` as Scoped once. 
DryIoc will automatically prefer a scoped dependency over singleton if its is injected scoped https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/RulesAndDefaultConventions.md#implicit-registration-selection-based-on-scope

There is one caveat though, because the implementation is planned to be changed based on the dynamic condition (being injected in scope or in the root) you need to setup both of your settings as `setup: Setup.With(asResolutionCall: true)` https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/RulesAndDefaultConventions.md#injecting-dependency-asresolutioncall

**Comment by publee** [comment]
All the troubles with child container and when I finally made it working, you show me that I don't need it :)
It's working finally and I am not going to touch it. But thank you for your help, your advices will be usefull for the next project.

**Comment by dadhi** [comment]
Child containers are very complex beasts because everyone wants their own breed. 
Glad that you won't need one.

**Comment by dadhi** [comment]
I've added `CreateChild` container which solves the original issue. Will be available in the v4.7.0
